[DRAFT] Fix: first push of pypi lesson (#164)
pyOpenSci/python-package-guide: tutorials/publish-pypi.md

> # Publish your Python package to PyPI

Author: This is the hatch lesson! i left a few TODO's at the top - the biggest one is 

should we teach pipx to install hatch? (it seems that would make sense) except pipx is a bit weird to install on windows potentially.

Reviewer: I am +1 on pipx. Even though the installation is not the easiest on Windows, I found it far easier to introduce tooling to beginners. With pipx, they do not mess with their global install, and they do not have any issue with tools being installed in virtual environment (and forgetting to activate said envs).

Reviewer: Oh, I just found out that pipx is available on scoop: https://scoop.sh/#/apps?q=pipx.
Then it is just as easy to install on Windows as on Linux/MacOS

Author: ok scoop is easy? i just wasn't sure.  i installed with homebrew @Batalex and it was really easy but you have to have homebrew setup of course. i also read that you can install pipx with pip but that hurts my brain a bit. i'm not sure where it would get installed if you did that.

:::{todo}
* emphasize that we recommended the trusted publisher GitHub action for most maintainers
* Make sure they add /dist to their .gitignore file. We have not discussed github workflows anywhere yet. Where does that fit?
* https://hatch.pypa.io/latest/intro/#existing-project <- hatch will migrate from setup.py for you - if we go with hatch then we may want to add this to the installable code lesson
* Should we install hatch with pipx?

```bash

> pipx install hatch

Author: i hear you. it sounds like we have mixed opinions about using pipx vs pip. it is very true that the `pipx` install will add cognitive load to this lesson. it's not the simplest altho you can pip install pipx but i could see that also confusing users as it kind of confuses me to think about it! 

maybe we can talk more in slack. i know some really like pipx and i do understand why. but on the other hand, most beginners won't likely be dealing with a ton of different package environments as they are just learning so keeping it simpler for now could be good. we could have a tiny breakout about pipx so they know it exists if they want to use it.

  installed package hatch 1.9.1, installed using Python 3.12.1
  These apps are now globally available
    - hatch
done! ✨ 🌟 ✨
```

:::

In the previous Python packaging lessons, you've learned:

1. What a Python package is
1. How to make your code installable.

:::{admonition} Learning Objectives
:class: tip

In this lesson you will learn how to:

> - Build your package's source (sdist) and wheel distributions

Reviewer 2: it might be helpful to either link out to a definition, or provide a short definition for sdist and wheel distributions here, and why they're important to python packaging

Author: we do have pages devoted to those terms and i do link to them in some places but don't want to link everywhere. Perhaps this might be a good place to test out our glossary for just these terms? i can create a tiny new pr to test it out as a demo!

Author: great -[ i created a draft pr here](https://github.com/pyOpenSci/python-package-guide/pull/167/files) that demos how the glossary works. So let's move this PR forward as is and we can add a glossary in a new pr. either you can create a totally new pr or you can build off the one that i created! because you are a maintainer you can push commits to my open pr. but because it's so tiny and unuseful as is - you can also just create a fresh PR if that sounds more straight forward!

> - Setup an account on testPyPI (the process is similar for the real PyPI)

Reviewer 3: ```suggestion
- Setup an account on test PyPI (the process is similar for the real PyPI)
```
Trying to find some consistency between all the testPyPI, test PyPI, test-PyPI, (test)PyPI 😃

Author: good point!! i'll do another find and replace check once i merge these commits!! ✨

> - Publish your package to testPyPI

Reviewer 3: ```suggestion
- Publish your package to test PyPI
```

You will do all of your development work in this lesson using Hatch.

Once your package is on PyPI you can publish it to the conda-forge channel of conda

> using [grayskull](https://conda.github.io/grayskull/).

Reviewer 2: this might just be a me being new to Python packaging thing, but there are soooo many names! I don't know that this lesson is the right place to solve that, but it does make me think that a glossary that we publish and can link to would be a huge help. 

I'm happy to take that on if you want, @lwasser!

Author: @kierisi see my comment above! we are on the SAME PAGE! how about this. i can add the infra for the sphinx glossary to get this started and then you can add terms? i'll just start with adding sdist and wheel as terms. it can be a separate pr. how does that sound?

Reviewer 2: @lwasser that would be great!

Author: excellent. i'm excited to try it out so i'll try to create the PR today as it should be really fast. then we can add links to bigger sections maybe in the glossary itself so i don't have to worry as much about crosslinking AND the user won't have to click around as much.

> You will learn how to publish to conda-forge in a future lesson.

Reviewer 2: I'm a bit unclear (again, being new to Python packaging) how publishing to the conda-forge channel of conda is different from publishing to conda-forge

Author: i clarified this locally but explaining what conda-forge is.

You need to build your Python package in order to publish it to PyPI (or Conda). The build process organizes your code and metadata into a distribution format that can be uploaded to PyPI and subsequently downloaded and installed by users.
:::

## Test PyPI.org vs PyPI

There are two "warehouses" that you can use to publish
your Python package.

1. **[Test PyPI](https://test.pypi.org):** Test PyPI is a version of the PyPI repository that can be used for testing. This is a great place to practice and learn how to publish a package without taking up space on the real PyPI servers.
2. **[Real PyPI](https://pypi.org):** This is the PyPI "warehouse" where you can officially publish your Python package. IMPORTANT: only publish your package to PyPI when you are ready for it to be used by others and/or confident that it will become a package that you maintain. PyPI is not a place to practice learning how to publish a Python package.

The steps for publishing on test PyPI vs. real PyPI are the same with the

> exception of a different url. Thus, in this lesson you will use testPyPI

Reviewer 3: ```suggestion
exception of a different url. Thus, in this lesson you will use test PyPI
```

to practice and learn.

> :::{todo}

Author: thank you - removed!

When this lesson is published -
in xx lesson, you will learn how to setup an automated release workflow on GitHub
using GitHub actions that will automate the PyPI publication process whenever
you create a new software release.
:::

## 4 Steps for publishing a Python package on PyPI

There are 4 things that you need to do to publish your Python package
to PyPI. You need to:

1. **Create a package development environment.** You will do this using Hatch.

> 1. [**Build your package**](../package-structure-code/python-package-distribution-files-sdist-wheel). Building a package is the process of turning your code into 2 distribution files: an sdist and a wheel. The wheel distribution file is particularly important for users who will `pip install` your package.

Reviewer 2: ```suggestion
1. [**Build your package**](../package-structure-code/python-package-distribution-files-sdist-wheel). Building a package is the process of turning your code into two distribution files: an sdist and a wheel. The wheel distribution file is particularly important for users who will `pip install` your package.
```

super small nit, and may be a style difference, but I've always gone by spelling out numbers between 0 and 9.

Author: excellent. i thijnk we had too many changes and i can no longer merge this but i did fix it locally! so i'm going to resolve. many thanks! i'm fairly certain we can setup a style guide enforcement set of rules using Vale (which @Batalex suggested in slack). i'm still playing with it but will see if we could create rules around numbers and such too.

```

Then view all of the current environments that hatch has access to:

```bash
> hatch env show
     Standalone
┏━━━━━━━━━┳━━━━━━━━━┓
┃ Name    ┃ Type    ┃
┡━━━━━━━━━╇━━━━━━━━━┩
│ default │ virtual │
└─────────┴─────────┘
```

Then activate the environment. Note that when you call a shell from a
Hatch environment, it will automatically install your package into the environment in development or editable mode.

```
# Hatch shell can be used to activate your environment
> hatch shell
... Installing project in development mode
source "/Path/to/env/here/hatch/env/virtual/pyosPackage/Mk7F5Y0T/sphinx-2i2c-theme/bin/activate"
```

View what's in the environment:

```bash
➜ pip list
Package         Version      Editable project location
--------------- ------------ ----------------------------------------------------
numpy           1.26.3
pandas          2.1.4
pip             23.3.1
pyosPackage     0.1.0        /path/to/your/project/here/pyosPackage
python-dateutil 2.8.2
pytz            2023.3.post1
six             1.16.0
tzdata          2023.4
```

At any time you can exit the environment using `exit`

> or <kbd>ctrl+d</kbd> on your keyboard.

Author: ok i wondered. so just keep `exit` as the command to return to the default envt? thank you for testing this! windows is a big concern of mine as i don't have a windows installation anymore to test on.

```
# Can someone please check this for Windows OS?

> > exit

Author: THANK YOU !!!

by default it uses venv[^venv].

hatch will:
1. Creates a new virtualenv (venv). That is located on your computer. You can customize the location of this environment if you wish  ....<where??>
2. Installs your package into the environment in editable mode (similar to `pip install -e`). This means it installs both your project and your project's dependencies as declared in your pyproject.toml file.

## Step 2: Build your package's sdist and wheel distributions

Once you have your development environment setup, you are ready to build your package using Hatch. Remember that building is the process of turning your Python package files into two distribution files:

1. The [wheel distribution](#python-wheel) is a pre-built version of your package. It useful for users as it can be directly installed using a tool such as `pip`. This file has the extension `.whl`.
2. The [source distribution](#python-source-distribution) is the files that make up your package in an unbuilt format. This file will have the extension `.tar.gz`.

You will use Hatch as a **Front end** tool that builds
your package's sdist and wheel using the [hatchling](https://hatch.pypa.io/latest/) build back-end.

> The hatchlin build back-end is used because you declared it in your pyproject.toml file in the [previous lesson](1-installable-code).

Reviewer: ```suggestion
The hatchling build back-end is used because you declared it in your pyproject.toml file in the [previous lesson](1-installable-code).
```

To build your package run:

```bash
➜ hatch build
──────────────── sdist ─────────────────
dist/pyospackage-0.1.0.tar.gz
──────────────── wheel ─────────────────
dist/pyospackage-0.1.0-py3-none-any.whl

```

:::{admonition} Learn more about the building a Python package
:class: tip
You can learn more about
building in the [build page of our packaging guide](../package-structure-code/python-package-distribution-files-sdist-wheel).
:::

:::{todo}
The sdist is important if you wish to [publish
your package to conda-forge](7-publish-conda-forge) which you will learn about in a later lesson.
:::

> :::{todo}

Author: Ahhhh so i'm trying to keep lessons that i have written but aren't online yet (and hence will create broken links for the users) in {todos}. i'm doing that because i want to embed the links throughout but if i just delete them, i'll have to remember where they go in a month or two when that lesson is live. so the todo sphinx extension allows me to create an admonition box that gets hidden when the page is rendered. then i can search for all of those and remove the TODO once the link is live.

i hope that makes sense. i couldn't think of a better way to do that and help my future self re-add those links back! i'll add a note in my next PR about this as well so it's less confusing!

➜ hatch build
────────────────────────────────────── sdist ──────────────────────────────────────
dist/pyospackage-0.1.0.tar.gz
────────────────────────────────────── wheel ──────────────────────────────────────
dist/pyospackage-0.1.0-py3-none-any.whl
:::

### <i class="fa-solid fa-wand-magic-sparkles"></i> Congratulations - you've created your Python package distribution files <i class="fa-solid fa-wand-magic-sparkles"></i>

You've now built your Python package and created your package distribution files. The next step is to setup
your account on PyPI so you can publish your package.

> ## Step 3. Setup your testPyPI account

Reviewer 3: ```suggestion
## Step 3. Setup your test PyPI account
```

> Next, you'll setup an account on `testPyPI`. Remember that you

Reviewer 3: ```suggestion
Next, you'll setup an account on test PyPI. Remember that you
```

> are using testPyPI here instead of the real PyPI as a way to

Reviewer 3: ```suggestion
are using test PyPI here instead of the real PyPI as a way to
```

safely learn how to publish a package without stressing the

> real PyPI's servers.

Reviewer 2: I'm wondering if "real" is the right term here, or if production would make more sense (to differentiate test from production). it's semantic, but I'm getting hung up on the idea that the test servers aren't "real"

Author: i've removed real through out and just used PyPI vs Test PyPI

:::

1. [Open up a web browser and go to the test PyPI website](https://test.pypi.org/).
2. [Create an account](https://test.pypi.org/account/register/) if you don't already have one. Be sure to store your password in a safe place!
3. Once you have an account setup, login to it.
4. Search on [https://test.pypi.org/](https://test.pypi.org/) (or pypi.org) to ensure that the package name that you have selected doesn't already exist. If you are using our test pyosPackage, then we suggest that you add your name or GitHub username to the end of the package name to ensure it's unique.

Example: `pyosPackage_yourNameHere`.

:::{todo}
Show them how to do this

1. update the project-name in the pyproject.toml file
2. update the module repo directory to be the same
:::

:::{figure-md} build-workflow-tutorial
<img src="../images/tutorials/testpypi-search.png" alt="This is a screenshot of the test PyPI website. At the top in the search bar, you can see the search for pyosPackage. The search return says  there were no results for pyosPackage Did you mean probpackage" width="700px">

> Before you try to upload to (test) PyPI, check to see if the name of your package is already taken. You can do that using

Reviewer 3: ```suggestion
Before you try to upload to test PyPI, check to see if the name of your package is already taken. You can do that using
```

> the search box at the top of the (test) PyPI website.

Reviewer 3: ```suggestion
the search box at the top of the test PyPI website.
```

:::

:::{admonition} Setup 2-factor (2FA) authentication

2-factor authentication is a secure login process that allows you to
use a backup device that only you can access to validate that the person logging in is really you. It addresses the issue of password phishing where someone else gains access to a password and can login to your account.

This matters on PyPI because someone could login to your account and upload a version of your package that has security issues. These issues will then impact all of your users when they download and install that version of the package.

> While you don't have to setup 2-factor authentication, we strongly

Reviewer: 2FA is now required since Jan 1st

Author: fixed! great catch!

suggest that you do so.
:::

## Step 4. Create a package upload token

To upload your package to PyPI, you will need to create a token. Ideally
this token is specific to the package that you are publishing.

However, if your package isn’t already on PyPI, then you will need to create a token for your account first and then create a package-specific token.

:::{admonition} Why create package-specific tokens?

It's ideal to create a package-specific token. When you create an account wide token this allows anyone with access to then access all of your PyPI projects. By creating a package specific token, you are limiting the scope of the token to only your specific package. This is just a safe way to set things up for you particularly if you are collaborating with others on package development.
:::

### Follow the steps below to create your token.

> * Login to testPyPI and go to your account settings

Reviewer 3: ```suggestion
* Login to test PyPI and go to your account settings
```

* Click on the **Add API Token** button
     * If you are new to using PyPI and don't have any packages there yet, OR if you have other packages on PyPI but are uploading a new package, you will need to create an account-wide token.
* When you create your token, be sure to copy the token value and store it in a secure place before closing that browser.

### Upload to PyPI using Hatch

Once you have the token in a safe place, you are ready to publish to
PyPI.

1. Set your username to `__token__`
1. Set your password to the token value, including the `pypi-` prefix

Your token should look something like this:

`pypi-abunchofrandomcharactershere...`

1. Finally run `hatch publish -r test`

-r stands for repository. In this case because you are publishing to test-PyPI you will use `-r test`. Hatch will then ask for a username and credentials.

* Add the word `__token__` for your username.
* Paste your PyPI token value in for the credential values.

```bash
❯ hatch publish -r test
Enter your username: __token__
Enter your credentials: <past-your-token-value-here>
dist/pyospackage-0.1.0-py3-none-any.whl ... already exists
dist/pyospackage-0.1.0.tar.gz ... already exists

> ```

Author: so here is a question - i see where you're going here!! 

is `hatch publish -r test -u __token__` 

any different from 

`hatch publish -r test`

and then the user enters __token__ as the username and the token value following the prompts?

i'm just thinking through whether a shorter command and then following the prompts might be easier for a user to remember vs a longer command. 

at the same time if hatch remembers it once the user calls the command that is also really nice. but agin the command 

`hatch publish -r test` would be the same there just would be no prompts after so that would be more consistent for a new user to remember. 

we can ask @ofek if we need `initialize-auth` for hatch to remember the authentication information. i am testing locally and it seems to remember the auth information by default.

> ```

Author: it looks to me like [things are saved](https://hatch.pypa.io/1.9/publish/#authentication) by default but perhaps i'm reading this wrong.

> ```

Author: `--initialize-auth -u __token__ ` did not ask me for a new token value. i'm not sure how to provide hatch with a new token value -- yet! we could also ask @flying-sheep :)

> ```

Reviewer 4: You don't need to initialize, it seems automatically as you mentioned. That flag was a feature request someone had to set up their build environment so nothing was required later.

You can set a new token by setting the environment variable of the command line flag but there probably should be a flag to force reset and ask again. Feel free to open a feature request!

> ```

Author: thank you @ofek !! i'll open an issue about this. at a minimum we can document it / perhaps update that page to be a bit more clear!

> ```

Reviewer 3: Okay, this all makes sense. To answer your first question @lwasser there is no difference between using `-u __token__` or entering it at the prompt. It was just my suggestion as `__token__` will be the same for all our readers, so the whole command can be copy/pasted and less need for interactive prompts. The token is of course different and secrets generally shouldn't be given as args anyways, so I left that part interactive even though there is an option to pass it in as well.
@ofek are credentials saved per-reposiroty and so changing `-r` will pick up new creds, or do we need to remove the test.pypi.org token to publish to pypi.org?

> ```

Reviewer 4: I believe per repository is how I cache but I would have to look at the code, also there is a feature request for making it even more granular per repository

If your credentials are value, Hatch will publish your package to test-PyPI.

Hatch also has a caching system so once you enter your credentials it will remember them.

> ## Install your package from TestPyPI

Reviewer 3: ```suggestion
## Install your package from test PyPI
```

Once your package upload is complete, you can install it from

> testPYPI. You can find the installation instructions on the testPyPI

Reviewer 3: ```suggestion
test PYPI. You can find the installation instructions on the test PyPI
```

landing page for your newly uploaded package.

:::{figure-md} testpypi-landing-page

> <img src="../images/tutorials/test-pypi-package.png" alt="A screenshot of the testPyPI page for pyosPackage. It says pyosPackage 0.1.0 at the top with the pip install instructions below. The landing page of the package has information from the package's readme file. " width="700px">

Reviewer 3: ```suggestion
<img src="../images/tutorials/test-pypi-package.png" alt="A screenshot of the test PyPI page for pyosPackage. It says pyosPackage 0.1.0 at the top with the pip install instructions below. The landing page of the package has information from the package's readme file. " width="700px">
```

> This is an example landing page for the pyosPackage that was just uploaded. Notice at the top of the page there is instructions for how to install the package from test PyPI. You can simply copy that code and use it to install your package from testPyPi locally.

Reviewer 3: ```suggestion
This is an example landing page for the pyosPackage that was just uploaded. Notice at the top of the page there is instruction for how to install the package from test PyPI. You can simply copy that code and use it to install your package from testPyPi locally.
```

:::

> As an example, [check out our pyOpenSci pyosPackage landing page on testPyPI](https://test.pypi.org/project/pyosPackage/). Notice the page has information about the current package version and also installation instructions as follows:

Reviewer 3: ```suggestion
As an example, [check out our pyOpenSci pyosPackage landing page on test PyPI](https://test.pypi.org/project/pyosPackage/). Notice the page has information about the current package version and also installation instructions as follows:
```

> `pip install -i https://test.pypi.org/simple/ pyosPackage`

Reviewer 3: Should probably note that this command _can_ install the newly published package, but that installing from test.pypi.org is generally considered a security risk. It might also fail to install even in cases where the same package would work coming from pypi.org

Author: @ucodery i'm curious - why is it considered a security risk to install from test pypi?

Author: i did make a text change to clarify this however. i left the security risk part out and just kept it general for the time being!

Reviewer 3: Because the authorship of packages is not the same; its first-come first-served for both repositories, and owning the name in one does not give you the name in the other. So by installing your own package pip also resolves all your dependencies, _all from test.pypi.org_, and installs those. There is no guarantee that the installed packages have the same owners, are available at the same versions, or even that the versions that do exist correspond to those on pypi.org.

### Time to install your package

* On your computer, activate the development environment that
you wish to install your newly published package in.
* Run the installation instructions for your package from test PyPI.

::::{tab-set}

:::{tab-item} Conda

```bash
> conda activate pyospkg-dev
> pip install -i https://test.pypi.org/simple/ youPackageNameHere
> conda list
```
:::

:::{tab-item} venv mac / Linux

> ```bash
> > source pyospackage-dev /bin/activate
> > pip install -i https://test.pypi.org/simple/ youPackageNameHere
> > pip list
> ```
> :::
> 
> :::{tab-item} venv windows
> ```bash
> > source pyospackage-dev /bin/activate
> > pip install -i https://test.pypi.org/simple/ youPackageNameHere
> > pip list
> ```

Reviewer 3: ```suggestion
```bash
> hatch shell
> pip install -i https://test.pypi.org/simple/ youPackageNameHere
> pip list
```
```

I'm not sure if this is specifically what we want to show. We already created the Hatch default env and editable-installed this package there, so installing over top of that from test.pypi.org seems wrong?

Author: oh you're right. that's really confusing isn't it? hatch by default i think installs your package in editable mode when it creates an environment (i'm still learning environments with hatch but that is what i took away from it). but i suppose instead of editable it would be installed fully into the environment. i'll definitely merge this edit but we may need to think that piece through as well.

 -->

## Package-specific token vs trusted publisher

For long run maintenance of your package, you have two options
related to PyPI publication.

1. You can create a package-specific token which you will use to publish your package (manually) to PyPI. This is a great option if you don't wish to automate your PyPI publication workflow.
2. You can also create an automated publication workflow on GitHub using GitHub actions. This is a great way to make the publication process easier and it also supports a growing maintainer team. In this case we suggest you don't worry about the token and instead setup a specific GitHub action that publishes your package when you make a release. You can then create a "trusted publisher" workflow on PyPI.

You will learn how to create the automated trusted publisher workflow in a followup lesson.

### OPTIONAL: If you want to use a manual token-based publication workflow

If you plan to use your token regularly to publish to PyPI, we strongly recommend going through the above steps again to create
a token specific to your new package.

To do this:

> 1. Go to testPyPI

Reviewer 3: ```suggestion
1. Go to test PyPI
```

1. Navigate to the "Your Projects" section of your account

> 2. Click on the manage button for the project that ou wish to add a token for

Reviewer 3: ```suggestion
2. Click on the manage button for the project that you wish to add a token for
```

3. Go to settings
4. Click on "Create a token for your-package-name-here"
5. Create the token and follow the steps above to open up the `.pypirc` file

> 6. Replace the old account wide token with your new package token.

Reviewer 3: This is the twine way. I don't know how to replace a token with Hatch. Maybe publish an existing version (so nothing actually uploads) but specify the new `--auth`?

Author: ahhh you're right. i'll go through and clean this up. i''ll also test to see if i use --auth if that is a way to force update the token value. leaving this unresolved until i can test it!

And you're all done!

## You have published your package to (test) PyPI!

> Congratulations. You have now successfully published your package to testPyPI. If you have a package that is ready for real-world use on the real pyPi, then you can follow the same steps to publish it on PyPI.org .

Reviewer 3: ```suggestion
Congratulations. You have now successfully published your package to test PyPI. If you have a package that is ready for real-world use on the real PyPI, then you can follow the same steps to publish it on PyPI.org.
```

# Publish your Python package to PyPI

<!-- TODOs:

* mention twine keyring - - https://twine.readthedocs.io/en/stable/#keyring-support
* emphasize that we recommended the trusted publisher github action for most maintainers

TODO: Go back to pyproj toml lesson and add a dev dependencies section with build and twine so they can just pip install -e .[dev] into their environment and be done

`pip install build`
-->

<!--
### Lesson TODOs ###

TODO: don't forget to have them add /dist to their .gitignore ...

https://hatch.pypa.io/latest/intro/#existing-project <- hatch will migrate from setup.py for you >
-->

In the previous Python packaging lessons, you've learned:

1. What a Python package is
1. How to make your code installable.

:::{admonition} Learning Objectives
:class: tip

In this lesson you will learn how to:

- Build your package's source (sdist) and wheel distributions
- Setup an account on testPyPI (the process is similar for the real PyPI)

> - Publish your package to testPyPI using [twine](https://twine.readthedocs.io/en/stable/)

Author: We could use hatch here instead

```
hatch publish test 
```

this will not only publish to test pypi but it will also ask for authentication and store it for easy user access

The steps for publishing on test PyPI vs. real PyPI are the same with the
exception of a different url. Thus, in this lesson you will use testPyPI
to practice and learn.

:::{todo}
When this lesson is published -
in xx lesson, you will learn how to setup an automated release workflow on GitHub
using GitHub actions that will automate the PyPI publication process whenever
you create a new software release.
:::

## 4 Steps for publishing a Python package on PyPI

There are 4 things that you need to do to publish your Python package
to PyPI. You need to:

1. Create a development environment
1. [**Build your package**](../package-structure-code/python-package-distribution-files-sdist-wheel). Building a package is the process of turning your code into 2 distribution files: an sdist and a wheel. The wheel distribution file is particularly important for users who will `pip install` your package.
2. **Create an account on (test) PyPI**: You will need to create a PyPI account and associated token which provides permissions for you to upload your package.

> 3. **Publish to PyPI using `twine`**: Once you have completed the above two steps, you are ready to use `twine` to publish your package!

Author: using `hatch publish test`

After that you can opt to create a automated workflow that publishes an updated
version of your package to PyPI every time you create a release on GitHub.

:::{admonition} Learn more about building Python packages in our guide
:class: tip
- [Learn more about building here](../package-structure-code/python-package-distribution-files-sdist-wheel)
- [Learn more about package distribution file that PyPI needs called the wheel](#python-wheel)
- [Learn more about the package distribution file that conda-forge will need on PyPI called the sdist (source distribution)](#python-source-distribution)

:::

## Step 1: Create a Python package development environment

The first step in building your package is to create a development environment that you
can work in. The Python environment will contain all of the dependencies needed to both install and work on your package.

:::{todo} This is with hatch

I don't think there is any additional information needed to use hatch here... to just build the package they won't need a dev environment i think they can use the default??

```bash
# This will create a default envt with your package installed in editable mode

> hatch env create

Author: I'm struggling with hatch environments. 

you can create an environment that is defined in the pyporject toml 

```
# Hatch config 
[tool.hatch.envs.dev]
dependencies = [
  "pytest"
]
```

but i don't understand how to actually enter / use the environment. so for instance if i want to run tests and pytest is in the dev envt how do i access that envt?

Author: ```
➜ hatch -e dev pytest
Usage: hatch [OPTIONS] COMMAND [ARGS]...
Try 'hatch -h' for help.

Error: No such command 'pytest'.

pyosPackage (☊ main) [✎ ×1 ] is 📦 v0.1.0 🐍 v3.12.1 
❯ hatch -e dev pytest
```

Author: `hatch run -e dev pytest`

```

:::

> 1. First add development dependencies in the `[optional.dependencies]` table of your `pyproject.toml` file. [Development dependencies](/package-structure-code/declare-dependencies.html) are dependencies that are only needed if someone wants to develop / work on your package. A user will never need development dependencies.

Author: First create a hatch config

```
# Hatch config 
[tool.hatch.envs.dev]
dependencies = [
  "pytest"
]
```

```toml
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "pyospackage"  # rename this if you plan to publish to test PyPI
# Here you add the package version manually.
# You will learn how

[project.optional-dependencies]
dev = [
  "twine",
  "build",
]
```

2. Create a Python package development environment if it is not already active. This development environment is where you will test out installation of your package. And also where you will develop your Python package.

If you don't already have a pyos-package development environment, create one now.

> ::::{tab-set}

Author: All of this could go away potentially with hatch.

:::{tab-item} Create Environment With Conda

```bash
conda create --name pyos-env python=3.10
conda activate pyos-env
```
:::

:::{tab-item} Create Environment With venv

```bash
# Create an environment on mac/linux
python -m venv pyos-pkg
# Activate the environment
source pyos-pkg/bin/activate
```

```bash
# Create the environment on Windows (i need a Windows tester)
python -m venv pyos-pkg
# Activate the environment on windows
pyos-pkg/bin/activate
```
::::

2. Once your new environment is created, install your package requirements from your `pyproject.toml` file.

:::{todo}
It probably makes sense to have them add an optional.dependencies section to their basic pyproject.toml file first. But this structure will  change if i decide to use hatch.
:::

```bash
> pip install -e ".[dev]"
```

You can check to see if twine has been installed using:

```bash
> pip list
...
traitlets                     5.9.0
twine                         4.0.2
...
```

> ## Step 2: Build your package's sdist and wheel distributions

Author: `hatch build`

You can learn more about both of these distribution files and package
building in general in the [build page of our packaging guide](../package-structure-code/python-package-distribution-files-sdist-wheel).
:::

The above two files are what you will need to publish your package to
PyPI. The `tar.gz` distribution file contains your source distribution (sdist). The sdist is important if you wish to publish
your package to conda-forge which you will learn about in a later lesson.

:::{todo}
The sdist is important if you wish to [publish
your package to conda-forge](7-publish-conda-forge) which you will learn about in a later lesson.
:::

:::{todo}
➜ hatch build
────────────────────────────────────── sdist ──────────────────────────────────────
dist/pyospackage-0.1.0.tar.gz
────────────────────────────────────── wheel ──────────────────────────────────────
dist/pyospackage-0.1.0-py3-none-any.whl
:::

### <i class="fa-solid fa-wand-magic-sparkles"></i> Congratulations - you've created your Python package distribution files <i class="fa-solid fa-wand-magic-sparkles"></i>

You've now created your package distribution. You're officially on
your way to publishing your package on PyPI.

> ## Step 3. Setup your testPyPI account

Author: So with hatch i think the above would become MUCH shorter like atleast 50% simpler. this section would stay as is to help them setup pypi.

:::

## Create a secure PyPI token

Follow the steps below to create your token.

* First login and go to your account settings in testPyPI
* Scroll down to the **API tokens** section
* Click on the **Add API Token** button
* If you are new to using PyPI and don't have any packages there yet, OR if you have other packages on PyPI but are uploading a new package, you will need to create an account-wide token.

:::{admonition} Tool options for publishing to PyPI

In this lesson you use twine to upload to PyPI. Note that other tools such as Flit, Hatch and PDM use slightly different approaches to storing and accessing your token to support uploading your package to PyPI. All of those tools are great ways to publish your package and provide a more end-to-end packaging solution.
:::

### Using a PyPI token

To use your API token, you do the following:

* Set your username to `__token__`
* Set your password to the token value, including the pypi- prefix

> ### Upload to PyPI using twine

Author: `hatch publish`
